pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/LightningSimulator.cpp

```diff
@@ -0,0 +1,563 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "LightningSimulator.hpp"
+
+#include "AdjointJacobianLQubit.hpp"
+#include "JacobianData.hpp"
+#include "LinearAlgebra.hpp"
+#include "MeasurementsLQubit.hpp"
+
+namespace Catalyst::Runtime::Simulator {
+
+auto LightningSimulator::AllocateQubit() -> QubitIdType {
+    size_t sv_id = this->device_sv->allocateWire();
+    return this->qubit_manager.Allocate(sv_id);
+}
+
+auto LightningSimulator::AllocateQubits(size_t num_qubits)
+    -> std::vector<QubitIdType> {
+    if (!num_qubits) {
+        return {};
+    }
+
+    // at the first call when num_qubits == 0
+    if (!this->GetNumQubits()) {
+        this->device_sv = std::make_unique<StateVectorT>(num_qubits);
+        return this->qubit_manager.AllocateRange(0, num_qubits);
+    }
+
+    std::vector<QubitIdType> result(num_qubits);
+    std::generate_n(result.begin(), num_qubits,
+                    [this]() { return AllocateQubit(); });
+    return result;
+}
+
+void LightningSimulator::ReleaseAllQubits() {
+    this->device_sv->clearData();
+    this->qubit_manager.ReleaseAll();
+}
+
+void LightningSimulator::ReleaseQubit(QubitIdType q) {
+    if (this->qubit_manager.isValidQubitId(q)) {
+        this->device_sv->releaseWire(this->qubit_manager.getDeviceId(q));
+    }
+    this->qubit_manager.Release(q);
+}
+
+auto LightningSimulator::GetNumQubits() const -> size_t {
+    return this->device_sv->getNumQubits();
+}
+
+void LightningSimulator::StartTapeRecording() {
+    RT_FAIL_IF(this->tape_recording, "Cannot re-activate the cache manager");
+    this->tape_recording = true;
+    this->cache_manager.Reset();
+}
+
+void LightningSimulator::StopTapeRecording() {
+    RT_FAIL_IF(!this->tape_recording,
+               "Cannot stop an already stopped cache manager");
+    this->tape_recording = false;
+}
+
+auto LightningSimulator::CacheManagerInfo()
+    -> std::tuple<size_t, size_t, size_t, std::vector<std::string>,
+                  std::vector<ObsIdType>> {
+    return {this->cache_manager.getNumOperations(),
+            this->cache_manager.getNumObservables(),
+            this->cache_manager.getNumParams(),
+            this->cache_manager.getOperationsNames(),
+            this->cache_manager.getObservablesKeys()};
+}
+
+void LightningSimulator::SetDeviceShots(size_t shots) {
+    this->device_shots = shots;
+}
+
+auto LightningSimulator::GetDeviceShots() const -> size_t {
+    return this->device_shots;
+}
+
+void LightningSimulator::SetDevicePRNG(std::mt19937 *gen) { this->gen = gen; }
```

Comment:
Just an aside, we shouldn't be using raw pointers in this way, as we are implicitly assuming the existence of `gen` and assuming the created class can also have ownership of it. We should use a `shared_ptr` resource instead to guarantee it will persist as long as required by instances of this class.

As this is likely a wider change across the Catalyst runtime, and here, I think it is best to mark this as a high priority fix to put onto the backlog to avoid potential sigsev issues.

Reply:
it is present in the LK plugin as well: https://github.com/PennyLaneAI/pennylane-lightning/blob/a55a58e5908550f3c31954a560ea080af04a9d7f/pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/LightningKokkosSimulator.cpp#L109

Reply:
@mlxd I suggest we first finish moving lightning.qubit's catalyst API to lightning repo proper, then factor out and clean up common code across lighting.qubit/kokkos/gpu, which will all be in lightning repo.

Reply:
Yep, onto the backlog, but with a high priority to fix :+1:

Reply:
Backlog it is